### PR TITLE
Add maildir backend `--colon` option

### PR DESCRIPTION
**Pull request by BoniLindsley**
Usage example:

```console
pymap --no-service admin --port 1143 maildir --colon '!' pymap_user.txt
```

Things I am unsure about:

- Are you okay with the option being called `--colon`? It is technically a delimiter.
- The use of `@property(pymap.backend.maildir.Config.colon)`. 
- How unit test would be done for this.

closes: #124

### CODE REVIEW

File: pymap/backend/maildir/__init__.py [hunk 1]
```diff
@@ -67,6 +67,8 @@ def add_subparser(cls, name: str, subparsers: Any) -> ArgumentParser:
                             help='maximum number of IO workers')
         parser.add_argument('--layout', metavar='TYPE', default='++',
                             help='maildir directory layout')
+        parser.add_argument('--colon', metavar='CHAR', default=':',
```
Comment (icgood) [inline pymap/backend/maildir/__init__.py:70]: Can you change this to `default=None` here? And then of course change the typing from `str` to `Optional[str]` and add an `if ... is not None` check before assigning to `maildir.colon` below. It doesn't make any functional difference, but it just keeps the scope of regression a bit smaller.

**Review by icgood — COMMENTED** [review]
>Things I am unsure about:
>
>  * Are you okay with the option being called --colon? It is technically a delimiter.

I think `--colon` is good, since it corresponds to the `Maildir.colon` attribute it's assigning to in the standard library.

>  * The use of @property(pymap.backend.maildir.Config.colon).

Yeah, it fits the pattern of the other arguments :+1:

>  * How unit test would be done for this.

Not (currently) possible. The test suite currently tests the in-memory backend only. To test the maildir and redis backends, I use [imaptest](https://imapwiki.org/ImapTest/) in the [docker container](https://hub.docker.com/repository/docker/icgood/imaptest) I put together.

```bash
# (on macOS)
docker run -it icgood/imaptest \
    host=host.docker.internal port=1143 user=demouser pass=demopass
```

**Comment by BoniLindsley** [comment]
I have not been able run pymap myself ever since the bump to 3.10 for various reasons, but hopefully the code is fine. The requested changes should all be there.

Since we are in 3.10, we can use `str | None` instead of `Optional[str]`. I stuck to the latter, assuming you prefer it.

And I finally realised that the `@property` on `colon` is possibly used to make it read-only. I keep forgetting that that is a use of `@property`.

**Review by icgood — APPROVED** [review]

**Comment by icgood** [comment]
> I have not been able run pymap myself ever since the bump to 3.10 for various reasons, but hopefully the code is fine. The requested changes should all be there.

Yeah, to be honest I haven't tested the maildir backend lately. I use the redis backend in my own setup. I'll give the maildir backend a shot this weekend and make sure it's still happy.

> Since we are in 3.10, we can use `str | None` instead of `Optional[str]`. I stuck to the latter, assuming you prefer it.

True. My instinct is to prefer `Optional[str]` but I'm curious if the community at large adopts the switch.
 
> And I finally realised that the `@property` on `colon` is possibly used to make it read-only. I keep forgetting that that is a use of `@property`.

Yes definitely that's one reason, but the _main_ reason is because `@property` with docstrings produces [cleaner docs](https://icgood.github.io/pymap/pymap.backend.maildir.html#pymap.backend.maildir.Config.layout) than using something like this:

```python
        #: A foo object.
        self.foo: Final = bar
```

I guess I also just don't really like the `#:` pseudo-docstring syntax, so for properties that I consider part of the "public" documented API I try and use `@property`.

**Comment by icgood** [comment]
Thanks for this PR!

Let me know if it works for you. For what it's worth, I have done **zero** Windows testing with pymap so there may be other issues. Perhaps it may be worth adding windows platform test runs to the [python-package.yml](https://github.com/icgood/pymap/blob/main/.github/workflows/python-package.yml) builds?

**Comment by BoniLindsley** [comment]
I am getting some TLS errors, so I cannot test this yet. That's probably a problem on my side, and unrelated to this issue.

On Windows, pymap fails trying to import `proxyprotocol`. Specifically, `AttributeError: module 'socket' has no attribute 'AF_UNIX'` in `proxyprotocol\v2.py`. That will not be a problem for me. I just need the maildir files to be accessible / moveable on Windows for archive sync-ing purposes.

**Pull request merged**